src/engine.rs

```diff
@@ -308,11 +311,95 @@ where
                 vec.push(read_entry_from_file::<M, _>(self.pipe_log.as_ref(), i)?);
             }
             ENGINE_READ_ENTRY_COUNT_HISTOGRAM.observe(ents_idx.len() as f64);
+            println!("[fetch_entries_to] time cost: {:?} us", start.elapsed().as_micros());
+            return Ok(ents_idx.len());
+        }
+        Ok(0)
+    }
+
+    pub fn fetch_entries_to_aio<M: Message + MessageExt<Entry = M> >(
+        &self,
+        region_id: u64,
+        begin: u64,
+        end: u64,
+        max_size: Option<usize>,
+        vec: &mut Vec<M::Entry>,
+    ) -> Result<usize> {
+        let start = Instant::now();
+        println!("[fetch_entries_to_aio] region id: {} left: {}, right: {}",region_id,begin,end);
+        let _t = StopWatch::new(&*ENGINE_READ_ENTRY_DURATION_HISTOGRAM);
+        if let Some(memtable) = self.memtables.get(region_id) {
+            let length = (end - begin) as usize;
+            let mut ents_idx: Vec<EntryIndex> = Vec::with_capacity(length);
+            memtable
+                .read()
+                .fetch_entries_to(begin, end, max_size, &mut ents_idx)?;
+            println!("[fetch_entries_to_aio] (stage1) time cost: {:?} us", start.elapsed().as_micros());
+
+            let mut new_block_flags:Vec<bool> = Vec::with_capacity(length);
+            let mut block_sum = 0;
+            for (t,i) in ents_idx.iter().enumerate(){
+                if match t {
+                    0 => true,
+                    _ => ents_idx[t-1].entries.unwrap() != ents_idx[t].entries.unwrap(),
+                }{
+                    block_sum += 1;
+                    new_block_flags.push(true);
+                }else{
+                    new_block_flags.push(false);
+                }
+            }
+            let mut a_list: Vec<aiocb> = Vec::with_capacity(block_sum);
```

Comment:
Hide aiocb inside `AioContext` (it holds `Vec<aiocb>` and create new ones when a new read request is issued via `fs::read_async`). No one other than the `DefaultFileSystem` should have access to aiocb.

Reply:
Do you mean to use one AioContext to manage all read requests? Then I need to redesign the AioContext, like this
`pub struct AioContext{
    fd_vec: Vec<Arc<LogFd>>,
    aio_vec: Vec<aiocb>,
    buf_vec: Vec<Vec<u8>>,
}`

Reply:
I'm not sure if there's a performance difference between waiting on each aiocb separately, or waiting all of them in one single syscall (maybe you can benchmark it too).

If there's no difference, then it makes sense to create new AioContext for each request. But either way, all the details must be put inside the file system implementation. E.g.
```
impl AioContext {
  pub fn wait() -> Result<()>;
  // UB if `wait()` is not called and returns `Ok(())`.
  pub fn data() -> &[u8];
}
```